concepts.md

```diff
@@ -91,6 +91,150 @@ An implementation may formally validate the structure of a CWL document using
 SALAD schemas located at
 https://github.com/common-workflow-language/common-workflow-language/tree/master/v1.1.0-dev1
 
+### map
+
+Note: This section is non-normative.
+> type: array&lt;ComplexType&gt; |  
+> map&lt;`key_field`, ComplexType&gt;
+
+The above syntax in the CWL specifications means there are two or more ways to write the given value.
+
+Option one is a array and is the most verbose option.
+In our example here we use the generic`ComplexType`, but
+in reality it would be one of `InputRecordField`, `OutputRecordField`,
+`CommandInputRecordField`, `SoftwarePackage`, `CommandInputParameter`,
+`CommandOutputParamter`, `EnvironmentDef`, `WorkflowStepInput`,
+`WorkflowInputParameter`, `WorkflowOutputParameter`, `WorkflowStep`,
+`ExpressionToolOutputParameter`, or a specific `*Requirement` entry.
+
+A generic example:
+```
+some_cwl_field:
+  - key_field: a_complex_type1
+    field2: foo
+    field3: bar
+  - key_field: a_complex_type2
+    field2: foo2
+    field3: bar2
+  - key_field: a_complex_type3
+```
+
+A specific example using [Workflow](Workflow.html#Workflow).[inputs](Workflow.html#WorkflowInputParameter):
+> array&lt;InputParameter&gt; |  
+> map&lt;`id`, `type` | InputParameter&gt;
+
+
+```
+inputs:
+  - id: workflow_input01
+    type: string
+  - id: workflow_input02
+    type: File
+    format: http://edamontology.org/format_2572
+  - id: workflow_input03
```

Comment:
The 3rd one is illegal, can't have inputs without a type

Reply:
Not according to https://www.commonwl.org/v1.1.0-dev1/Workflow.html#WorkflowInputParameter "required: false"